File: backend/routes/notifications.py
from flask import Blueprint, request, jsonify
from flask_security import login_required, current_user
from database import db
from models import Notifications, Tasks, Projects, TeamMembers
from deadline_warnings import DeadlineWarningEngine
from datetime import datetime, timedelta

notifications_bp = Blueprint('notifications', __name__)

# ---------- GET USER NOTIFICATIONS ----------
@notifications_bp.route("/", methods=["GET"])
@login_required
def get_user_notifications():
    """Get notifications for the current user with filtering options"""
    
    # Query parameters for filtering
    unread_only = request.args.get("unread_only", "false").lower() == "true"
    notification_type = request.args.get("type")  # deadline_warning, task_overdue, general
    priority = request.args.get("priority")  # low, medium, high, critical
    project_id = request.args.get("project_id")
    limit = min(int(request.args.get("limit", 50)), 100)  # Cap at 100
    offset = int(request.args.get("offset", 0))

    # Base query
    query = Notifications.query.filter_by(user_id=current_user.id)

    # Apply filters
    if unread_only:
        query = query.filter(Notifications.is_read == False)
    
    if notification_type:
        query = query.filter(Notifications.type == notification_type)
    
    if priority:
        query = query.filter(Notifications.priority == priority)
    
    if project_id:
        try:
            project_id_int = int(project_id)
            # Verify user has access to this project
            team_membership = TeamMembers.query.filter_by(
                project_id=project_id_int, 
                user_id=current_user.id
            ).first()
            if team_membership:
                query = query.filter(Notifications.project_id == project_id_int)
            else:
                return jsonify({"error": "Project not found or user not authorized"}), 403
        except ValueError:
            return jsonify({"error": "Invalid project_id"}), 400

    # Get total count for pagination
    total_count = query.count()
    
    # Apply pagination and ordering
    notifications = query.order_by(
        Notifications.created_at.desc()
    ).offset(offset).limit(limit).all()

    # Format response
    notification_data = []
    for notification in notifications:
        data = {
            "id": notification.id,
            "title": notification.title,
            "message": notification.message,
            "type": notification.type,
            "priority": notification.priority,
            "is_read": notification.is_read,
            "created_at": notification.created_at.isoformat(),
            "task_id": notification.task_id,
            "project_id": notification.project_id
        }
        
        # Add related task/project info if available
        if notification.task:
            data["task_title"] = notification.task.title
            data["task_status"] = notification.task.custom_status.name if notification.task.custom_status else None
        
        if notification.project:
            data["project_name"] = notification.project.name
            
        notification_data.append(data)

    return jsonify({
        "notifications": notification_data,
        "total_count": total_count,
        "unread_count": Notifications.query.filter_by(
            user_id=current_user.id, 
            is_read=False
        ).count(),
        "has_more": offset + limit < total_count
    }), 200


# ---------- MARK NOTIFICATION AS READ ----------
@notifications_bp.route("/<int:notification_id>/read", methods=["PUT"])
@login_required
def mark_notification_read(notification_id):
    """Mark a specific notification as read"""
    
    notification = Notifications.query.filter_by(
        id=notification_id, 
        user_id=current_user.id
    ).first()
    
    if not notification:
        return jsonify({"error": "Notification not found"}), 404
    
    notification.is_read = True
    db.session.commit()
    
    return jsonify({"message": "Notification marked as read"}), 200


# ---------- MARK ALL NOTIFICATIONS AS READ ----------
@notifications_bp.route("/read-all", methods=["PUT"])
@login_required
def mark_all_notifications_read():
    """Mark all notifications as read for the current user"""
    
    notification_type = request.args.get("type")  # Optional: only mark specific type as read
    
    query = Notifications.query.filter_by(user_id=current_user.id, is_read=False)
    
    if notification_type:
        query = query.filter(Notifications.type == notification_type)
    
    updated_count = query.update({"is_read": True})
    db.session.commit()
    
    return jsonify({
        "message": f"Marked {updated_count} notifications as read"
    }), 200


# ---------- DELETE NOTIFICATION ----------
@notifications_bp.route("/<int:notification_id>", methods=["DELETE"])
@login_required
def delete_notification(notification_id):
    """Delete a specific notification"""
    
    notification = Notifications.query.filter_by(
        id=notification_id, 
        user_id=current_user.id
    ).first()
    
    if not notification:
        return jsonify({"error": "Notification not found"}), 404
    
    db.session.delete(notification)
    db.session.commit()
    
    return jsonify({"message": "Notification deleted"}), 200


# ---------- DEADLINE WARNING ANALYSIS ----------
@notifications_bp.route("/deadline-analysis", methods=["POST"])
@login_required
def run_deadline_analysis():
    """Trigger deadline warning analysis for all projects the user has access to"""
    
    # Get all projects user is a member of
    user_projects = db.session.query(Projects).join(TeamMembers).filter(
        TeamMembers.user_id == current_user.id
    ).all()
    
    if not user_projects:
        return jsonify({"message": "No projects found for user"}), 200
    
    # Run analysis
    try:
        analysis_summary = DeadlineWarningEngine.analyze_all_tasks()
        
        return jsonify({
            "message": "Deadline analysis completed",
            "summary": analysis_summary,
            "analyzed_projects": len(user_projects)
        }), 200
        
    except Exception as e:
        return jsonify({"error": f"Analysis failed: {str(e)}"}), 500


# ---------- TASK DEADLINE INSIGHTS ----------
@notifications_bp.route("/task-insights/<int:task_id>", methods=["GET"])
@login_required
def get_task_deadline_insights(task_id):
    """Get detailed deadline insights for a specific task"""
    
    # Verify user has access to the task
    task = Tasks.query.get(task_id)
    if not task:
        return jsonify({"error": "Task not found"}), 404
    
    # Check if user is a team member of the project
    team_membership = TeamMembers.query.filter_by(
        project_id=task.project_id, 
        user_id=current_user.id
    ).first()
    
    if not team_membership:
        return jsonify({"error": "Task not found or user not authorized"}), 403
    
    try:
        insights = DeadlineWarningEngine.get_task_deadline_insights(task_id)
        return jsonify(insights), 200
        
    except Exception as e:
        return jsonify({"error": f"Failed to get insights: {str(e)}"}), 500


# ---------- NOTIFICATION PREFERENCES ----------
@notifications_bp.route("/preferences", methods=["GET"])
@login_required
def get_notification_preferences():
    """Get notification preferences for the current user"""
    # This is a placeholder for future user preference customization
    # For now, return default preferences
    
    default_preferences = {
        "deadline_warnings": {
            "enabled": True,
            "frequency": {
                "critical": 6,  # hours
                "high": 12,
                "medium": 24
            }
        },
        "email_notifications": False,  # Future feature
        "push_notifications": True    # Future feature
    }
    
    return jsonify(default_preferences), 200


# ---------- NOTIFICATION STATISTICS ----------
@notifications_bp.route("/stats", methods=["GET"])
@login_required
def get_notification_stats():
    """Get notification statistics for the current user"""
    
    today = datetime.utcnow().date()
    week_ago = today - timedelta(days=7)
    month_ago = today - timedelta(days=30)
    
    stats = {
        "total_notifications": Notifications.query.filter_by(user_id=current_user.id).count(),
        "unread_notifications": Notifications.query.filter_by(
            user_id=current_user.id, 
            is_read=False
        ).count(),
        "this_week": Notifications.query.filter(
            Notifications.user_id == current_user.id,
            Notifications.created_at >= week_ago
        ).count(),
        "this_month": Notifications.query.filter(
            Notifications.user_id == current_user.id,
            Notifications.created_at >= month_ago
        ).count(),
        "by_type": {},
        "by_priority": {}
    }
    
    # Get breakdown by type
    type_counts = db.session.query(
        Notifications.type, 
        db.func.count(Notifications.id)
    ).filter_by(user_id=current_user.id).group_by(Notifications.type).all()
    
    for notification_type, count in type_counts:
        stats["by_type"][notification_type] = count
    
    # Get breakdown by priority
    priority_counts = db.session.query(
        Notifications.priority, 
        db.func.count(Notifications.id)
    ).filter_by(user_id=current_user.id).group_by(Notifications.priority).all()
    
    for priority, count in priority_counts:
        stats["by_priority"][priority] = count
    
    return jsonify(stats), 200 

File: synergy-sphere-frontend/src/store/notifications.js
import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import axios from '../axios'

export const useNotificationsStore = defineStore('notifications', () => {
  // State
  const notifications = ref([])
  const unreadCount = computed(() => notifications.value.filter(n => !n.is_read).length)
  const loading = ref(false)
  const error = ref(null)
  
  // Computed
  const unreadNotifications = computed(() => 
    notifications.value.filter(n => !n.is_read)
  )
  
  const deadlineWarnings = computed(() => 
    notifications.value.filter(n => n.type === 'deadline_warning')
  )
  
  const criticalNotifications = computed(() => 
    notifications.value.filter(n => n.priority === 'critical')
  )

  // Actions
  async function fetchNotifications(options = {}) {
    try {
      // Build query parameters from options
      const params = new URLSearchParams()
      
      if (options.limit) params.append('limit', options.limit)
      if (options.offset) params.append('offset', options.offset)
      if (options.unread_only) params.append('unread_only', 'true')
      if (options.type) params.append('type', options.type)
      if (options.priority) params.append('priority', options.priority)
      if (options.project_id) params.append('project_id', options.project_id)
      
      const queryString = params.toString()
      const url = queryString ? `/notifications?${queryString}` : '/notifications'
      
      const response = await axios.get(url)
      
      // Handle the API response format which wraps notifications in a data object
      if (response.data.notifications) {
        notifications.value = response.data.notifications
      } else {
        notifications.value = response.data
      }
    } catch (error) {
      console.error('Failed to fetch notifications:', error)
    }
  }
  
  // Mark notification as read
  async function markAsRead(notificationId) {
    try {
      await axios.put(`/notifications/${notificationId}/read`)
      const notification = notifications.value.find(n => n.id === notificationId)
      if (notification) {
        notification.is_read = true
      }
    } catch (error) {
      console.error('Failed to mark notification as read:', error)
    }
  }
  
  // Get deadline insights for a task
  async function getTaskDeadlineInsights(taskId) {
    try {
      const response = await axios.get(`/notifications/task-insights/${taskId}`)
      return response.data
    } catch (error) {
      console.error('Failed to get task deadline insights:', error)
      return null
    }
  }
  
  // Clear all notifications
  async function clearAll() {
    try {
      await axios.delete('/notifications')
      notifications.value = []
    } catch (error) {
      console.error('Failed to clear notifications:', error)
    }
  }

  async function markAllAsRead(type = null) {
    try {
      const params = type ? `?type=${type}` : ''
      const response = await axios.put(`/notifications/read-all${params}`)
      
      // Update local state
      notifications.value.forEach(notification => {
        if (!type || notification.type === type) {
          notification.is_read = true
        }
      })
      
      if (!type) {
        unreadCount.value = 0
      } else {
        unreadCount.value = notifications.value.filter(n => !n.is_read).length
      }
      
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Failed to mark notifications as read'
      console.error('Mark all as read error:', err)
      return false
    }
  }

  async function deleteNotification(notificationId) {
    try {
      await axios.delete(`/notifications/${notificationId}`)
      
      // Update local state
      const index = notifications.value.findIndex(n => n.id === notificationId)
      if (index !== -1) {
        const notification = notifications.value[index]
        if (!notification.is_read) {
          unreadCount.value = Math.max(0, unreadCount.value - 1)
        }
        notifications.value.splice(index, 1)
      }
      
      return true
    } catch (err) {
      error.value = err.response?.data?.error || 'Failed to delete notification'
      console.error('Delete notification error:', err)
      return false
    }
  }

  async function runDeadlineAnalysis() {
    loading.value = true
    error.value = null
    
    try {
      const response = await axios.post('/notifications/deadline-analysis')
      
      // Refresh notifications after analysis
      await fetchNotifications()
      
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Failed to run deadline analysis'
      console.error('Deadline analysis error:', err)
      return false
    } finally {
      loading.value = false
    }
  }

  async function getTaskInsights(taskId) {
    try {
      const response = await axios.get(`/notifications/task-insights/${taskId}`)
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Failed to get task insights'
      console.error('Task insights error:', err)
      return null
    }
  }

  async function getNotificationStats() {
    try {
      const response = await axios.get('/notifications/stats')
      return response.data
    } catch (err) {
      error.value = err.response?.data?.error || 'Failed to get notification stats'
      console.error('Notification stats error:', err)
      return null
    }
  }

  function addNotification(notification) {
    // Add new notification to the beginning of the array
    notifications.value.unshift(notification)
    if (!notification.is_read) {
      unreadCount.value += 1
    }
  }

  function clearError() {
    error.value = null
  }

  function reset() {
    notifications.value = []
    unreadCount.value = 0
    loading.value = false
    error.value = null
  }

  return {
    // State
    notifications,
    unreadCount,
    loading,
    error,
    
    // Computed
    unreadNotifications,
    deadlineWarnings,
    criticalNotifications,
    
    // Actions
    fetchNotifications,
    markAsRead,
    markAllAsRead,
    deleteNotification,
    runDeadlineAnalysis,
    getTaskInsights,
    getNotificationStats,
    addNotification,
    clearError,
    reset,
    getTaskDeadlineInsights,
    clearAll
  }
}) 